includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp-cached-data.php

```diff
@@ -0,0 +1,349 @@
+ *
+ * The purpose of this class is to implement a non-obstrusive chat, in which refreshing the cache will happen in the background in an async request
+ * and will never keep the user waiting.
+ *
+ * 1. It will check for the information in cache (favor wp_cache because it's faster than options, especially with memcached)
+ * 2. If it does not exist, it will check for the last_cached value, stored as an option (with auto_load as false, to avoid unnecessary queries)
+ * 3. It will then dispatch an async request to refresh the cache, while returning the last cached data immediately
+ * 4. It will clear the last cached data, to make sure cache will be refreshed eventually, even if the async request fails for any reason
+ * 5. The async request will fetch data from the Mailchimp server and pouplate the cache
+ * 6. The first time we ask for data in a given list, there will be no cache nor last_cached option, so we'll fetch the data synchronously
+ */
+final class Newspack_Newsletters_Mailchimp_Cached_Data {
+
+	/**
+	 * The cache group name
+	 *
+	 * @var string
+	 */
+	const CACHE_GROUP = 'newspack_nl_mailchimp_data';
+
+	/**
+	 * The last cache option name
+	 *
+	 * @var string
+	 */
+	const OPTION_NAME = 'newspack_nl_mailchimp_last_cache';
+
+	/**
+	 * The ajax action name used to dispatch the cache refresh
+	 *
+	 * @var string
+	 */
+	const AJAX_ACTION = 'newspack_nl_mailchimp_refresh_cached_data';
+
+	/**
+	 * Memoized data to be served across the same request
+	 *
+	 * @var ?array
+	 */
+	private static $memoized_data;
+	
+	/**
+	 * Initializes this class
+	 */
+	public static function init() {
+		add_action( 'wp_ajax_' . self::AJAX_ACTION, [ __CLASS__, 'handle_dispatch_refresh' ] );
+	}
+
+	/**
+	 * Get segments of a given list
+	 *
+	 * @param string $list_id The List ID.
+	 * @throws Exception In case of errors while fetching data from the server.
+	 * @return array The list segments
+	 */
+	public static function get_segments( $list_id ) {
+		$data = self::get_data( $list_id );
+		return $data['segments'] ?? null;
+	}
+
+	/**
+	 * Get Interest Categories (aka Groups) of a given list
+	 *
+	 * @param string $list_id The List ID.
+	 * @throws Exception In case of errors while fetching data from the server.
+	 * @return array The list interest categories
+	 */
+	public static function get_interest_categories( $list_id ) {
+		$data = self::get_data( $list_id );
+		return $data['interest_categories'] ?? null;
+	}
+
+	/**
+	 * Get merge_fields of a given list
+	 *
+	 * @param string $list_id The List ID.
+	 * @throws Exception In case of errors while fetching data from the server.
+	 * @return array The list merge_fields
+	 */
+	public static function get_merge_fields( $list_id ) {
+		$data = self::get_data( $list_id );
+		return $data['merge_fields'] ?? null;
+	}
+
+	/**
+	 * Retrieves the main Mailchimp instance
+	 *
+	 * @return Newspack_Newsletters_Mailchimp
+	 */
+	private static function mc() {
```

Comment:
A more descriptive function name would make it more understandable :)

Reply:
changed